lib/ttfunk/encoded_string.rb

```diff
@@ -0,0 +1,97 @@
+require 'stringio'
+
+module TTFunk
+  class UnresolvedPlaceholderError < StandardError
+  end
+
+  class DuplicatePlaceholderError < StandardError
+  end
+
+  class EncodedString
+    def initialize
+      yield self if block_given?
+    end
+
+    def <<(obj)
+      case obj
+      when String
+        io << obj
+      when Placeholder
+        add_placeholder(obj)
+        io << "\0" * obj.length
+      when self.class
+        # adjust placeholders to be relative to the entire encoded string
+        obj.placeholders.each_pair do |_, placeholder|
+          add_placeholder(placeholder.dup, placeholder.position + io.length)
+        end
+
+        self << obj.unresolved_string
+      end
+
+      self
+    end
+
+    def align(width = 4)
+      self.class.new do |new_str|
+        new_str << self
+
+        if length % width > 0
+          new_str << "\0" * (width - length % width)
+        end
+      end
+    end
+
+    def length
+      io.length
+    end
+
+    def string
+      unless placeholders.empty?
+        raise UnresolvedPlaceholderError, 'string contains '\
+          "#{placeholders.size} unresolved placeholder(s)"
+      end
+
+      io.string
+    end
+
+    def bytes
+      string.bytes
+    end
+
+    def unresolved_string
```

Comment:
This and all the following public methods can be marked protected. They still will be accessible from withing this class but won't be a part of a public API.

Reply:
The `#unresolved_string` method is called several times outside the context of the `EncodedString` class, so it will need to remain public.